openmmmcmc/states.py

```diff
@@ -0,0 +1,1903 @@
+
+        Notes
+        -----
+        The reduced potential is defined as in Ref. [1]
+
+        u = \beta [U(x) + p V(x) + \mu N(x)]
+
+        where the thermodynamic parameters are
+
+        \beta = 1/(kB T) is the inverse temperature
+        p is the pressure
+        \mu is the chemical potential
+
+        and the configurational properties are
+
+        x the atomic positions
+        U(x) is the potential energy
+        V(x) is the instantaneous box volume
+        N(x) the numbers of various particle species (e.g. protons of
+             titratible groups)
+
+        References
+        ----------
+        [1] Shirts MR and Chodera JD. Statistically optimal analysis of
+        equilibrium states. J Chem Phys 129:124105, 2008.
+
+        Examples
+        --------
+
+        Compute the reduced potential of a water box at 298 K and 1 atm.
+
+        >>> from openmmtools import testsystems
+        >>> waterbox = testsystems.WaterBox(box_edge=20.0*unit.angstroms)
+        >>> system, positions = waterbox.system, waterbox.positions
+        >>> state = ThermodynamicState(system=waterbox.system,
+        ...                            temperature=298.0*unit.kelvin,
+        ...                            pressure=1.0*unit.atmosphere)
+        >>> integrator = openmm.VerletIntegrator(1.0*unit.femtosecond)
+        >>> context = state.create_context(integrator)
+        >>> context.setPositions(waterbox.positions)
+        >>> sampler_state = SamplerState.from_context(context)
+        >>> u = state.reduced_potential(sampler_state)
+
+        If the sampler state is incompatible, an error is raised
+
+        >>> incompatible_sampler_state = sampler_state[:-1]
+        >>> state.reduced_potential(incompatible_sampler_state)
+        Traceback (most recent call last):
+        ...
+        ThermodynamicsError: The sampler state has a different number of particles.
+
+        In case a cached SamplerState containing the potential energy
+        and the volume of the context is not available, the method
+        accepts a Context object and compute them with Context.getState().
+
+        >>> u = state.reduced_potential(context)
+
+        """
+        # Read Context/SamplerState n_particles, energy and volume.
+        if isinstance(context_state, openmm.Context):
+            n_particles = context_state.getSystem().getNumParticles()
+            openmm_state = context_state.getState(getEnergy=True)
+            potential_energy = openmm_state.getPotentialEnergy()
+            volume = openmm_state.getPeriodicBoxVolume()
+        else:
+            n_particles = context_state.n_particles
+            potential_energy = context_state.potential_energy
+            volume = context_state.volume
+
+        # Check compatibility.
+        if n_particles != self.n_particles:
+            raise ThermodynamicsError(ThermodynamicsError.INCOMPATIBLE_SAMPLER_STATE)
+
+        beta = 1.0 / (unit.BOLTZMANN_CONSTANT_kB * self.temperature)
```

Comment:
> why not make this a property

Sure! I can add this.

> Also can we add the `/ unit.AVOGADRO_CONSTANT_NA` to beta instead of reduced potential?

I think the thermodynamic beta doesn't have the Avogadro constant in it. Also, if I don't strip it from the reduced potential I need to multiply the pressure-volume work by `Na` too since it has incompatible units otherwise (it's a minor optimization).